nexus/db-queries/src/db/sec_store.rs

```diff
@@ -53,16 +56,46 @@ impl steno::SecStore for CockroachDbSecStore {
     }
 
     async fn record_event(&self, event: steno::SagaNodeEvent) {
-        debug!(&self.log, "recording saga event";
+        let log = self.log.new(o!(
             "saga_id" => event.saga_id.to_string(),
-            "node_id" => ?event.node_id,
-            "event_type" => ?event.event_type,
-        );
+            "node_id" => event.node_id.to_string(),
+            "event_type" => format!("{:?}", event.event_type),
+        ));
+
+        debug!(&log, "recording saga event");
         let our_event = db::saga_types::SagaNodeEvent::new(event, self.sec_id);
 
-        // TODO-robustness This should be wrapped with a retry loop rather than
-        // unwrapping the result.  See omicron#2416.
-        self.datastore.saga_create_event(&our_event).await.unwrap();
+        backoff::retry_notify_ext(
+            // This is an internal service query to CockroachDB.
+            backoff::retry_policy_internal_service(),
+            || {
+                // XXX: do we need to filter the kinds of errors we retry on?
+                // Are there particular database errors which are unrecoverable
+                // and where we should just panic.
```

Comment:
In this specific case, I *think* it's possible to get a `Error::ObjectAlreadyExists` error under certain conditions (because we're supplying the `ErrorHandler::Conflict` to `public_error_from_diesel` within `saga_create_event`.

I believe that all other errors are currently 500s, but that one seems permanent.

Reply:
I think what I'll do here is mark these errors in the logs as likely requiring operator intervention, while retaining the current behavior of retrying forever. That seems like a reasonable short-term path forward.

Longer term we may want to make this part of whatever operator alerting system we build.